### DISCUSSION TITLE
Delegate pattern for actions

### DISCUSSION THREAD

**Discussion by rcarver** [pointfreeco/swift-composable-architecture#1007] [Show and tell]
Here's an idea I've been enjoying lately, just sharing in case it's interesting to others – I've started using a kind of 'delegate' pattern to clarify which child actions are designed to be observed by a parent domain.

A common use case to close a sheet:
1. User taps button, triggering action `.tappedSaveButton`
2. The reducer decides to show a confirmation dialog
3. The confirmation dialog triggers actions `.tappedCancel` or `.tappedConfirm`

When this domain is composed, it gets confusing as to which action should be observed to handle tearing down the child. Consider as things evolve; perhaps the confirmation dialog is added later. There'd be nothing to tell users of this domain to change how they observe closing.

Instead, define a public 'delegate' interface that clarifies which actions should be observed, like this:

```swift
enum SheetAction {
    case tappedSaveButton
    case tappedCancel
    case tappedConfirm
    case delegate(DelegateAction)
    enum DelegateAction {
        case didClose(shouldSave: Bool)
    }
}
```

A few benefits I've found:
1. It clarifies which actions are internal vs external. Since all actions must be part of the public interface this gives you a way to define the external contract.
1. The domain's reducer can simply do `case .delegate:` and skip everything that it doesn't need to handle.
1. It's clearer if a parent is observing actions it shouldn't

Has anyone done something similar? Any better ways to approach this? Does the word 'delegate' capture the intent? 

---

You really don't need any tools to use this pattern, but here's a way to clean up triggering the delegating action.

```swift
public protocol DelegatingAction {
    associatedtype Action
    static func delegate(_ action: Action) -> Self
}

extension Effect where Output: DelegatingAction {
    static func delegate(_ action: Output.Action) -> Self {
        .init(value: .delegate(action))
    }
}
```

And a working example:

```swift
enum TestDelegating {

    struct CountState: Equatable {
        var count: Int = 0
    }

    enum CountAction: DelegatingAction {
        case increment
        case delegate(DelegateAction)
        enum DelegateAction {
            case didIncrementCount(Int)
        }
    }

    static let countReducer = Reducer<CountState, CountAction, Void>.combine(
        Reducer { state, action, _ in
            switch action {
            case .increment:
                state.count += 1
                return .delegate(.didIncrementCount(state.count))
            case .delegate:
                return .none
            }
        }
    )

    struct ParentState: Equatable {
        var count = CountState()
        var lastCountAt: Date?
    }

    enum ParentAction {
        case count(CountAction)
    }

    static let parentReducer = Reducer<ParentState, ParentAction, Void>.combine(
        countReducer.pullback(
            state: \.count,
            action: /ParentAction.count,
            environment: { $0 }
        ),
        Reducer { state, action, _ in
            switch action {
            case .count(.delegate(.didIncrementCount)):
                state.lastCountAt = Date()
                return .none
            case .count:
                return .none
            }
        }
    )

    struct IncrementButton: View {
        var store: Store<CountState, CountAction>
        var body: some View {
            WithViewStore(store) { viewStore in
                Button {
                    viewStore.send(.increment)
                } label: {
                    Text("Increment")
                }
            }
        }
    }
}

struct TestDelegating_Preview: PreviewProvider {
    static let store = Store<TestDelegating.ParentState, TestDelegating.ParentAction>(
        initialState: .init(),
        reducer: TestDelegating.parentReducer,
        environment: ()
    )
    static var previews: some View {
        VStack {
            TestDelegating.IncrementButton(
                store: store.scope(
                    state: \.count,
                    action: TestDelegating.ParentAction.count
                )
            )
            WithViewStore(store) { viewStore in
                Text(String(describing: viewStore.count))
                if let date = viewStore.lastCountAt {
                    Text(String(describing: date))
                }
            }
        }
    }
}
```

**Comment by tgrapperon** [comment]
Great topic @rcarver! I don't use this approach for delegates, but I've started breaking my actions with subtypes like you did with `DelegateAction`. For example, I'm usually using an inner enum to host embeddings of child actions, and it's indeed easy to "carpet-break" all embedded actions in one `case .child: break` when no additional logic is required. I guess the same approach can be used to segregate "user" actions coming from the UI, from "internal" actions for example.

The `Delegate` approach looks fine. It can seem a little "impure" for a child feature to acknowledge the existence of an "external" being, but this is not a problem per se. One functionally similar approach I'm using in a few places is to define "logic-less" child reducers that are only handling consistency of the `state`'s properties for the UI. These child reducers can then be pulled-back into reducers that are specializing them to a given context[^1]. This is in some way similar to the `Delegate` approach, albeit using composition instead of baking this degree of freedom in the child domain itself. 
[^1]: There's a `filter` higher order reducer in isowords that can mute some actions in a given reducer. This can be very convenient to contextualize reducers in some setups, but I've found easier to define "logic-less" reducers with the cases I had to deal with.

I guess the `Delegate` approach has its merits by limiting/declaring what's possible or not, and it may be an alternative to composition for exposing the API of features shared as third-party libraries. Maybe this can be standardized a little more: Delegates have very diverse functions in UIKit/AppKit: datasource, callbacks, observation, etc. Exposing an action doesn't immediately makes clear if someone has to handle it at all cost, if this is a signal of something happening in the child feature, or if this is some action we can send to control the child feature. Maybe some additional structure on this pattern could help clarifying things (but maybe I'm already bikeshedding).

But this is a very interesting topic!

**Reply by rcarver** [reply]
Great feedback, thanks @tgrapperon!

> The Delegate approach looks fine. It can seem a little "impure" for a child feature to acknowledge the existence of an "external" being, but this is not a problem per se. 

It's interesting that you say this, as modeling parent/child relationships and coordinating the responsibilities of each is something I'm struggling with a lot. Since composing domains generally means coordinating the actions of each, it seems like defining these contracts is at the core of good design with TCA. I'm specifically looking at this in terms of small SPA modules and taking inspiration from the [Modularization series from Pointfree](https://www.pointfree.co/episodes/ep172-modularization-part-2) and how they carved out specific integration points for the [ItemRowViewModel](https://github.com/pointfreeco/episode-code-samples/blob/e3d74fd92142882d408008b72656e201ce511050/0172-modularization-pt2/Inventory/Sources/ItemRowFeature/ItemRow.swift#L39-L40)

```swift
public var onDelete: () -> Void = {}
public var onDuplicate: (Item) -> Void = { _ in }
```

> One functionally similar approach I'm using in a few places is to define "logic-less" child reducers that are only handling consistency of the state's properties for the UI

Can you point me to an example of "logic-less"? I'm not quite getting it.

> Delegates have very diverse functions in UIKit/AppKit: datasource, callbacks, observation, etc. Exposing an action doesn't immediately makes clear if someone has to handle it at all cost, if this is a signal of something happening in the child feature, or if this is some action we can send to control the child feature. Maybe some additional structure on this pattern could help clarifying things (but maybe I'm already bikeshedding).

Yes 100%. Building on the example above, I chose the concept "delegate" since the most common case in UIKit is to notify of some event. That said, I think this pattern would probably work to support "sendable" delegate actions as well. Which itself is another side of this design discussion. 

---

Stepping back, this is really a design question along the lines of "how do you best define and document the public contract of a domain using TCA?"

---

For exploration, here's three ways of defining an actions enum with "internal" and "external" actions. You could swap those words for public, private, delegate, notify or any other term. It's interesting to look at the ergonomics of each of these, both from the domain's (internal) perspective as well as from the consumer (external) perspective. 

```swift
/// Internal actions are exposed directly, exteral actions are consolidated.
enum ExternalSheetAction {
    case tappedSaveButton
    case tappedCancel
    case tappedConfirm
    case external(ExternalAction)
    enum ExternalAction {
        case didClose(shouldSave: Bool)
    }
}
let externalReducer = Reducer<Void, ExternalSheetAction, Void> { _, action, _ in
    switch action {
    case .tappedSaveButton:
        // set state for confirmation dialog
        return .none
    case .tappedCancel:
        return Effect(value: .external(.didClose(shouldSave: false)))
    case .tappedConfirm:
        return Effect(value: .external(.didClose(shouldSave: true)))
    case .external:
        return .none
    }
}

enum ComposingExternalSheetAction {
    case sheet(ExternalSheetAction)
}
let composingExternalReducer = Reducer<Void, ComposingExternalSheetAction, Void> { _, action, _ in
    switch action {
    case .sheet(.external(.didClose)):
        // handle closing sheet
        return .none
    case .sheet:
        return .none
    }
}
```

```swift
/// Internal actions are consolidated, external actions are exposed directly.
enum InternalSheetAction {
    case `internal`(InternalAction)
    case didClose(shouldSave: Bool)
    enum InternalAction {
        case tappedSaveButton
        case tappedCancel
        case tappedConfirm
    }
}
let internalReducer = Reducer<Void, InternalSheetAction, Void> { _, action, _ in
    switch action {
    case .internal(.tappedSaveButton):
        // set state for confirmation dialog
        return .none
    case .internal(.tappedCancel):
        return Effect(value: .didClose(shouldSave: false))
    case .internal(.tappedConfirm):
        return Effect(value: .didClose(shouldSave: true))
    case .didClose:
        return .none
    }
}

enum ComposingInternalSheetAction {
    case sheet(InternalSheetAction)
}
let composingInternalReducer = Reducer<Void, ComposingInternalSheetAction, Void> { _, action, _ in
    switch action {
    case .sheet(.didClose):
        // handle closing sheet
        return .none
    case .sheet(.internal):
        return .none
    }
}
```

```swift
/// Both internal and external actions are consolidated
enum InternalExternalSheetAction {
    case `internal`(InternalAction)
    case external(ExternalAction)
    enum InternalAction {
        case tappedSaveButton
        case tappedCancel
        case tappedConfirm
    }
    enum ExternalAction {
        case didClose(shouldSave: Bool)
    }
}
let internalExternalReducer = Reducer<Void, InternalExternalSheetAction, Void> { _, action, _ in
    switch action {
    case .internal(.tappedSaveButton):
        // set state for confirmation dialog
        return .none
    case .internal(.tappedCancel):
        return Effect(value: .external(.didClose(shouldSave: false)))
    case .internal(.tappedConfirm):
        return Effect(value: .external(.didClose(shouldSave: true)))
    case .external:
        return .none
    }
}

enum ComposingInternalExternaSheetAction {
    case sheet(InternalExternalSheetAction)
}
let composingInternalExternaReducer = Reducer<Void, ComposingInternalExternaSheetAction, Void> { _, action, _ in
    switch action {
    case .sheet(.external(.didClose)):
        // handle closing sheet
        return .none
    case .sheet(.internal):
        return .none
    }
}
```

**Reply by tgrapperon** [reply]
> Can you point me to an example of "logic-less"? I'm not quite getting it.

Sorry, it wasn't very clear indeed. I should have used "internal-logic" instead to describe them. I'm thinking about handling things that are specifically internal to the reducer. Roughly how you would use a `@State` property in SwiftUI. For example, imagine a "keyboard" domain, where you have to maintain a certain internal logic (switching the keys content when "shift" is pressed, handling repeats when the key is kept pressed, etc.) but also communicate with the outside world: what happens when a key is pressed. This is only an illustrative example, but is shows how the Delegate/composition questions are similar.

Fortunately or not, TCA doesn't impose strong restrictions over the way Actions are arranged, and this degree of freedom can let some users undecided with the optimal approach to compose/arrange their domains. It seems clear that, as it evolves, TCA will go toward the more general / less constrained solution, so if some prescription or rule should be used, it will happen on the user's side. It doesn't mean that interesting pattern won't emerge though. I'll look more closely at your example later.

One thing I would like to experiment recently is trying to turn more SwiftUI ideas into TCA concepts. One candidate appears in the proto branch (see the discussion #832). We can imagine that, at some point, we'll build reducers using a `ReducerBuilder`. It may then become more natural to see a given reducer as a collection of specialized reducers. I don't think the current approach to modelize actions monolithically is satisfactory: in one switch, we handle many responsibilities: updating the UI, triggering background tasks, navigating, etc. We can already split reducers and actions with the tools we have, but it is a little cumbersome to pull them back. Hopefully, the `ReducerBuilder` could help streamlining this and democratizing the practice. It seems then natural that delegate-style callbacks could be handled by some kind of `DelegateReducer` working with its own set of actions. Once the `DelegateReducer` is a type, we can maybe start to encode behaviors in its properties/types.

**Comment by rcarver** [comment]
Here's another sketch of how a parent could observe a child's delegate actions more clearly.

```swift
enum CountAction: DelegatingAction {
    case increment
    case delegate(DelegateAction)
    enum DelegateAction {
        case didIncrementCount(Int)
    }
}

enum ParentAction {
    case count(CountAction)
}

static let parentReducer = Reducer<ParentState, ParentAction, Void>.combine(
    countReducer.pullback(
        state: \.count,
        action: /ParentAction.count,
        environment: { $0 }
    ),
    Reducer { state, action, _ in
        switch action {
        case .count:
            return .none
        }
    }
        .delegate(action: /ParentAction.count) { state, action, environment in
            switch action {
            case .didIncrementCount:
                state.lastCountAt = Date()
                return .none
            }
        }
)
```

Since there's nothing to observe outside the delegate it could be simplified further:

```swift
static let parentReducer = Reducer<ParentState, ParentAction, Void>.combine(
    countReducer.pullback(
        state: \.count,
        action: /ParentAction.count,
        environment: { $0 }
    )
        .delegate(action: /ParentAction.count) { state, action, environment in
            switch action {
            case .didIncrementCount:
                state.lastCountAt = Date()
                return .none
            }
        }
)
```
